### PR TITLE
Fix cgroups determination in exec implementation

### PR DESCRIPTION
**Pull request by YJDoc2**
This fixes the way we determine cgroup name as well as the way we set the cgroup in case of exec. Right now on the main, following would fail even though they should succeed

```
podman run -d --runtime /path/to/youki busybox sleep 10m
podman exec -it --runtime /path/to/youki SHA-HERE /bin/sh
```
This is because 
1. we copy over the namespaces from "main" container to tenant, but not the cgroups
2. In case of rootless, we directly use container id as cgroup name, but that fails when trying to decode the cgroup path

Thus, we fix it here by also copying over the cgroup config from init to tenant, and if cgroup path is not present using `:youki:id` instead .

I have also added a test for exec in the rootless tests we have.

~NOT READY FOR MERGE YET.~

### CODE REVIEW

**Comment by YJDoc2** [comment]
With this PR, the number of failed podman tests goes down to 105 from 157 : https://github.com/YJDoc2/youki/actions/runs/8138706191 :tada:

File: crates/libcgroups/src/systemd/manager.rs [hunk 1]
```diff
@@ -353,6 +353,10 @@ impl CgroupManager for Manager {
         if pid.as_raw() == -1 {
             return Ok(());
         }
+        if self.client.transient_unit_exists(&self.unit_name) {
+            tracing::debug!("Transient unit {:?} already exists", self.unit_name);
+            return Ok(());
+        }
```
Comment (YJDoc2) [inline crates/libcgroups/src/systemd/manager.rs:359]: I'm not sure if this  is correct. For exec, the unit will exist (as the container is running) , so we should probably add the pid of the new exec process into it? If so then in case of tenant container, we need a call to set unit props with pid modified? @utam0k may I ask you to take a look at this?

Comment (utam0k) [reply crates/libcgroups/src/systemd/manager.rs:359]: In terms of OCI Runtime spec, such a behavior isn't defined. I'm not an expert on systemd, but anyway, we have to put a tenant pid into a cgroup which is the same cgroup of running container pid.

**Comment by lengrongfu** [comment]
LGTM

**Review by lengrongfu — APPROVED** [review]

File: crates/libcgroups/src/systemd/manager.rs (hunk 1 above)
Comment (YJDoc2) [reply crates/libcgroups/src/systemd/manager.rs]: Hey, fixed this. Now the add_task will either create a new cgroup with the provided pid, or if cgroup already exists, it will add the pid into that. I have verified it using systemd cgroup impl, and changed for cgroups-c2 , but not sure how to verify that, as even docker uses systemd cgroups.

**Comment by YJDoc2** [comment]
@lengrongfu @utam0k May I ask one of you to take a look at the  newer changes I have pushed?

File: crates/libcontainer/src/container/tenant_builder.rs [hunk 2]
```diff
@@ -327,8 +327,16 @@ impl TenantContainerBuilder {
...
 
+        let linux = spec.linux().as_ref().unwrap();
+        let linux = if linux.cgroups_path().is_some() {
+            LinuxBuilder::default()
+                .namespaces(ns)
+                .cgroups_path(linux.cgroups_path().as_ref().unwrap().clone())
+                .build()?
+        } else {
+            LinuxBuilder::default().namespaces(ns).build()?
+        };
```
[the remark this answers is not in the document]
Comment (lengrongfu) [reply crates/libcontainer/src/container/tenant_builder.rs:339]:
```rust
        let linux_spec = spec.linux().as_ref().unwrap();
        let linux = if linux_spec.cgroups_path().is_some() {
            LinuxBuilder::default()
                .namespaces(ns)
                .cgroups_path(linux_spec.cgroups_path().as_ref().unwrap().clone())
                .build()?
        } else {
            LinuxBuilder::default().namespaces(ns).build()?
        };
```
Maybe it would be better if we rename this var name.

File: crates/libcgroups/src/systemd/manager.rs [hunk 3]
```diff
@@ -353,6 +353,11 @@ impl CgroupManager for Manager {
         if pid.as_raw() == -1 {
             return Ok(());
         }
+        if self.client.transient_unit_exists(&self.unit_name) {
```
Comment (utam0k) [inline crates/libcgroups/src/systemd/manager.rs:356]: Can we add a unit tet for this change?

**Review by utam0k — CHANGES_REQUESTED** [review]
I noticed this leads to breaking changes for existing users due to changing a default cgroup path. In my opinion, affecting rootless users can be allowed, but breaking the change for regular users(non-rootless) is more important. Can we avoid affecting regular users?

File: crates/libcontainer/src/container/tenant_builder.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (YJDoc2) [reply crates/libcontainer/src/container/tenant_builder.rs:339]: yep, will do that as well :+1:

**Comment by YJDoc2** [comment]
> I noticed this leads to breaking changes for existing users due to changing a default cgroup path. In my opinion, affecting rootless users can be allowed, but breaking the change for regular users(non-rootless) is more important. Can we avoid affecting regular users?

Hey, while this is breaking , I'm not sure which part specifically are you talking about with respect to default cgroup path. If it is this :  https://github.com/containers/youki/pull/2720/files#diff-36cd5e5a1222b5c819958e4045a242cbff00ae22a8a1699229f5c6220dafdb6bR150-R153

Then, this only changes the when the cgroup path was not set, and I had observed this behavior when creating a tenant container via exec. For the normal container, the path is correctly set and would not change in this. Also the previous behavior of using just the `container_id` as the path is somewhat wrong, as in a valid cgroup path check, we expect it to have two `:` indicating the parent and prefix [here](https://github.com/containers/youki/blob/main/crates/libcgroups/src/systemd/manager.rs#L77-L109) in the systemd manager, which using just the `container_id` is wrong (I have observed that docker and podman with root still use the systemd manager) .

If you meant something else, please correct me. Thanks!

File: crates/libcgroups/src/systemd/manager.rs (hunk 3 above)
Comment (YJDoc2) [reply crates/libcgroups/src/systemd/manager.rs:356]: ok, willl do :+1:

**Comment by utam0k** [comment]
> Hey, while this is breaking , I'm not sure which part specifically are you talking about with respect to default cgroup path. If it is this : https://github.com/containers/youki/pull/2720/files#diff-36cd5e5a1222b5c819958e4045a242cbff00ae22a8a1699229f5c6220dafdb6bR150-R153

Yes

> Then, this only changes the when the cgroup path was not set, and I had observed this behavior when creating a tenant container via exec. For the normal container, the path is correctly set and would not change in this. 

I wonder what happens in the following case:
1. a user runs a container(Container A) with no-set cgroup path with the latest youki version(`v0.3.2`)
2. a user updates the release version, including this PR
3. a user tries `exec` to Container A

In this above case, Every 1 and 3 processes find out different cgroup paths, right?

**Comment by YJDoc2** [comment]
Hey,

>I wonder what happens in the following case:
    a user runs a container(Container A) with no-set cgroup path with the latest youki version(v0.3.2)
    a user updates the release version, including this PR
    a user tries exec to Container A
In this above case, Every 1 and 3 processes find out different cgroup paths, right?

Note : below youki 3.2 = youki on tag v0.3.2 ; branch youki = youki compiled from this branch

So I tried this and when I launch container A (with sudo  pdoman) with youki 3.2 , then try to exec into it with youki 3.2, it gives error of
```
Error: OCI runtime error: thread 'main' panicked at crates/libcontainer/src/process/container_intermediate_process.rs:49:82:
called `Result::unwrap()` on an `Err` value: Systemd(CgroupsPath(MalformedPath("348136f0a314677e10ac6b3d09b2e7430a70ee6b2b01a23239f0bb4f18cfd150")))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
exec failed : failed to receive. "waiting for intermediate process". BrokenChannel
```
which is what I mentioned about the default cgroup path being incorrect in my previous comment.

However, if I try to launch container A (with sudo podman) youki 3.2 and exec into it using branch youki , it works due to the cgroups path being set correctly in this branch. Also, I verified that there were two processes in the cgroup of container A when I tried to exec into it using branch youki.

Similarly using branch youki for container A and exec works too. Trying to exec into container launched with branch youki using youki 3.2 also fails due to malformed path.

In short, the exec was broken for both rootful and rootless on v0.3.2 , so I think changing the cgroup path determination wouldn't cause any issues when using containers launched with youki 0.3.2 with branch youki.

Does this address your concern? Do you want me to try any other permutations?

**Comment by YJDoc2** [comment]
Also, I need to check this in detail, but I suspect that changes in the cgroup path determination don't get applied at all due to the changes in setting the cgroups  path [here](https://github.com/containers/youki/pull/2720/files#diff-f1b7c5673621e51900ef703849bf21a2b78474d14349f133a6c82e137a13fb92R331-R339) for exec containers. 

Previously we would not set that path, thus the path determination finding `None` and using `container_id` as path, which lead to the failure. However, after we set the cgroup path, it will always take that as the cgroup path for exec/tenant container. I need to see in which case does the cgroup path can be `None` (I think it can be case if someone is using youki directly withou docker/podman or they are doing some manual cgroup setup) .

**Comment by utam0k** [comment]
> I think it can be case if someone is using youki directly withou docker/podman or they are doing some manual cgroup setup

Yes, I imagined this situation(Sorry for my lack of words 🙏)
Is there any case at all where podman, contained, and CRI-O is not setting the cgroup path? I think it is possible to write a Migration Guide and get away with it.

**Comment by YJDoc2** [comment]
@utam0k I have added a test for systemd cgroup manager regarding the changes here. The CI is failing as the dbus pipe is not set up correctly, so I'll fix that up ; but can you take a look at the test I have added, and check if that approach is ok, or any changes are needed - commit : https://github.com/containers/youki/pull/2720/commits/4a952119fa0d1bf585d0202a4f1402b031075c59 ? Thanks :)

**Comment by utam0k** [comment]
The commit in itself looks acceptable to me. I wonder whether this PR should be written as a breaking change or not.

**Comment by YJDoc2** [comment]
Thanks, I'll work on fixing the CI then.

>The commit in itself looks acceptable to me. 

I'm not sure, while some of the internals have changed, the overall effect is making exec work which was not working previously, thus I added the bug label. I'll think more on this :+1:

> I wonder whether this PR should be written as a breaking change or not.

**Comment by utam0k** [comment]
LGTM except for the CI failed

**Comment by YJDoc2** [comment]
@utam0k please take a look. I have changed the approach of adding the task - before I wrote directly to the cgroup proc file, which could be incorrect / have permission errors ; Now I use the systemd dbus api to add the process via dbus call. I have also fixed the tests, although for the task addition test, I needed to add some more mounts and one hack script so that it can run properly in the cross container.

See these three commits for these new changes : https://github.com/containers/youki/pull/2720/commits/75c851b78d4dab8eb54e3cb868bcb1b271c98a7a https://github.com/containers/youki/pull/2720/commits/3fe7d04ae4fa173ccdf2ab342ab56ec09cfa4e4c https://github.com/containers/youki/pull/2720/commits/df5ebb29ef85b8cd8106227667ea244e38552aae

Thanks :)

**Review by utam0k — APPROVED** [review]
Depending on DBus is better than the previous approach, +1

**Comment by utam0k** [comment]
Thanks a lot, @YJDoc2
